### ISSUE TITLE
[java] EmptyFinalizer should not trigger if finalize method is final and class is not

### ISSUE THREAD

**Issue by kohlschuetter** [pmd/pmd#4742]
**Affects PMD Version:**  6.55.0

**Rule:** EmptyFinalizer

Please provide the rule name and a link to the rule documentation:
https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#emptyfinalizer

**Description:**

The rule declares "Empty finalize methods serve no purpose and should be removed. ".

This is not true for the case that the method is marked "final" in a class that is not marked "final".

SpotBugs recently introduced a check "CT_CONSTRUCTOR_THROW" that errs out if an exception is thrown in a non-final class's constructor, because nefarious subclasses may perform a so-called "Finalizer Attack". Also see https://wiki.sei.cmu.edu/confluence/display/java/OBJ11-J.+Be+wary+of+letting+constructors+throw+exceptions

Moreover, declaring an empty, final finalizer method may be a performance optimization, as per https://stuartmarks.wordpress.com/2022/04/27/why-write-an-empty-finalize-method/

**Code Sample demonstrating the issue:**

```java
package example;

public class PMDTest {
  protected final Object someState = initObject();

  public PMDTest() throws Exception {
  }

  @Override
  @SuppressWarnings("NoFinalizer" /* checkstyle */)
  @Deprecated
  protected final void finalize() {
  }

  private Object initObject() {
    return new Object();
  }
}
```

**Expected outcome:**

PMD reports a violation at line 12, but that's wrong. That's a false positive.

**Running PMD through:** *[Maven]*

**Comment by kohlschuetter** [comment]
OK, I take it back. I think the explanation of this warning should be improved, however.

It looks like with an empty finalizer method, we also prevent Cleaners from running properly. I'm not entirely sure why, but tests are failing sporadically in my project after adding these empty finalize blocks, even for code that has Cleaners enabled...

**Comment by Vampire** [comment]
This is still an issue with 7.18.0

**Comment by Vampire** [comment]
And the same should be considered for `EmptyMethodInAbstractClassShouldBeAbstract`.
If it overwrites a method and is declared final in a non-final class, it should be fine.

**Comment by mrclmh** [comment]
@Vampire Since this is a different rule, could you please create a new issue with a specific example for your use case? Thank you.

**Comment by Vampire** [comment]
Sure, makes sense.
Use-case is for example exactly the same, but should imho be considered for all cases not only `finalize`. :-)
#6279
